BRANCHING.md

```diff
@@ -0,0 +1,71 @@
+  - The difference between `develop` and `berkeley` is that `berkeley` will be the actual hardfork release, while `develop` is subsequent softfork release candidate, softfork after `berkeley`. `develop` is just not tested as rigorously, but it's softfork compatible with `berkeley`. So if `berkeley` can be thought of as 2.0, then `develop` is 2.01.
+- Direction of merge:
+  - The back-merging direction is `master` to `compatible` to `rampup` to the next hardfork (now, `berkeley`) to `develop`.
+    - So `develop` contains all the changes from the more stable branches.
+  - Merging forward (when introducing features/fixes) is up this stack, but the place the change is introduced can be different (e.g. `compatible` for softfork features, `develop` for more experimental/next release).
+- Hard fork/Release:
+  - Whenever a hard fork happens, the code in  `develop` is released.  When this happens, the current `compatible` is entirely discarded and a new `compatible` gets created based off of `develop`
+  - `release/1.X.X` branches are made off of `compatible` and tagged with alpha and beta tags until the code is deemed stable, then the `release/1.X.X` branch is merged into `master` and given a stable tag. Whenever code is tagged, if anything is missing in in the upstream branches (compatible, develop) then the tagged branch is also merged back for consistency.
+  - So after Berkeley release: `berkeley` branch will become the new `master`. `berkeley` will be removed from `proof-systems`. `develop` will be renamed into `compatible`.
+- `o1js-main`: compatible with testnet, but has latest `proof-systems` features so that they can be used in `o1js`
+  - Contains mina changes from `rampup`
+  - But `proof-systems/develop` which by default is used by `mina/develop`.
+  - Uses `o1js/main` and `o1js-bindings/main` as explained [here](https://github.com/o1-labs/o1js/blob/main/README-dev.md#branch-compatibility?).
+  - When `proof-systems/develop` is too cutting-edge and the adaptations of its changes haven't been landed in mina, `o1js` will use the `proof-systems/o1js-main` branch which is lagging behind `proof-systems/develop` a bit.
+
+
+### Diagram
+
+**TODO** improve diagram
+
+![git-flow_david-wong.png](./git-flow_david-wong.png)
+
+## Day to day: which branch should I use?
+
+When developing a feature, if it’s not something that breaks compatibility, then you should be developing a feature branch, called `foo_COMP` for example, based off of `compatible`.  If you’re not sure whether or not your changes are breaking, they probably are not and should build upon compatible.
```

Comment:
This really depends on what the feature/bug fix is scoped for. If the feature is scoped for HF and is not compatible against a running public testnet, then base it off of the HF branch (for example, `berkeley`). If it is a bug fix required for a public testnet testing upcoming hf then base it off of `rampup`. If a feature/enhancement/bug fix is scoped for a mainnet then base it off of  `compatible. I am not sure if it makes sense to have a general suggestion re which branch to use

Reply:
Added this to the text.